### PR TITLE
Adding modifications and refactors for production deployments

### PR DESCRIPTION
**Pull request by aric49**
Modifications and refactoring I have been doing to enhance resiliency and production readiness . Major changes are:

1) Refactoring DNS Resiliency to use self-hosted Kube-DNS instead of outside DNS service
2) Deploying Calico with BGP Peering
3) Validations to ensure pods are coming up correctly
4) Installing base packages and directory locations required for volume mounts
5) Basic installations for Helm/Tiller/MaaS (coming soon)

### CODE REVIEW

File: ansible/roles/deploy_helm/templates/kube-controller-manager.json.j2 [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      }
+    ],
+    "containers": [
+      {
+        "name": "kube-controller-manager",
+        "image": "quay.io/attcomdev/kube-controller-manager:{{ kube_controller_manager_version }}",
+        "command": [
+          "kube-controller-manager",
+          "--address=127.0.0.1",
+          "--leader-elect",
```
Comment (v1k0d3n) [inline ansible/roles/deploy_helm/templates/kube-controller-manager.json.j2:35]: i believe this needs to be `--leader-elect=true`

File: ansible/roles/bootstrap/tasks/main.yml [hunk 2]
```diff
@@ -1,5 +1,27 @@
...
+- name: Install base packages
+  apt:
+    name: "{{ item }}"
+    state: present
+  with_items:
+    - "docker.io"
+    - "htop"
+    - "vim"
+    - "ethtool"
+    - "tmux"
```
Comment (v1k0d3n) [inline ansible/roles/bootstrap/tasks/main.yml]: i think most developers would feel `tmux` is a requirement :)

but is it needed for the deployment?

File: ansible/roles/deploy_helm/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    path: "{{ item }}"
+    state: directory
+  with_items:
+    - "/var/lib/openstack-helm/ceph/osd"
+    - "/var/lib/openstack-helm/ceph/ceph"
+    - "/var/lib/openstack-helm/ceph/mon"
+    - "/var/lib/nova/instances"
+
+- name: Install Sigil for Ceph Secrets
+  shell: curl -L https://github.com/gliderlabs/sigil/releases/download/v0.4.0/sigil_0.4.0_Linux_x86_64.tgz | tar -zxC /usr/local/bin
```
Comment (v1k0d3n) [inline ansible/roles/deploy_helm/tasks/main.yml:30]: do we want to potentially make `0.4.0` a variable?

File: ansible/roles/deploy_helm/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+  delay: 15
+
+- name: Check for Kubernetes dashboard
+  shell: hyperkube kubectl get pods -o wide --all-namespaces | grep kubernetes-dashboard
+  register: dashboard_check
+  ignore_errors: true
+
+- name: Deploy Kubernetes Dashboard
+  shell: hyperkube kubectl create -f https://rawgit.com/kubernetes/dashboard/master/src/deploy/kubernetes-dashboard.yaml
+  when: dashboard_check | failed
```
Comment (v1k0d3n) [inline ansible/roles/deploy_helm/tasks/main.yml:62]: do you want dashboard to be an optional element long term as well, or did you want to keep it built into the deployment? may help to have a version included in defaults as well.

File: ansible/roles/deploy_helm/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+- name: Check if Helm is installed
+  stat:
+    path: /usr/local/bin/helm
+  register: helm_installed
+
+- name: Install helm
+  shell: curl https://raw.githubusercontent.com/kubernetes/helm/master/scripts/get > /root/get_helm.sh
+  when: helm_installed.stat.exists == False
```
Comment (v1k0d3n) [inline ansible/roles/deploy_helm/tasks/main.yml:71]: have you considered downloading targeting a specific version of helm (and adding to a defaults section)?

File: ansible/roles/kubelet/templates/calico.yaml [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
+           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
+    spec:
+      # Only run this pod on the master.
+      nodeSelector:
+        kubeadm.alpha.kubernetes.io/role: master
+      hostNetwork: true
+      containers:
+        - name: calico-etcd
+          image: gcr.io/google_containers/etcd:2.2.1
```
Comment (v1k0d3n) [inline ansible/roles/kubelet/templates/calico.yaml:79]: may want to version this?

File: ansible/roles/kubelet/templates/calico.yaml [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
+           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
+    spec:
+      hostNetwork: true
+      containers:
+        # Runs calico/node container on each Kubernetes node.  This
+        # container programs network policy and routes on each
+        # host.
+        - name: calico-node
+          image: quay.io/calico/node:v1.0.2
```
Comment (v1k0d3n) [inline ansible/roles/kubelet/templates/calico.yaml:149]: might want to consider versioning this as well.

File: ansible/roles/kubelet/templates/calico.yaml [hunk 8]
```diff
@@ -0,0 +1,322 @@
...
+            - mountPath: /lib/modules
+              name: lib-modules
+              readOnly: true
+            - mountPath: /var/run/calico
+              name: var-run-calico
+              readOnly: false
+        # This container installs the Calico CNI binaries
+        # and CNI network config file on each node.
+        - name: install-cni
+          image: calico/cni:v1.5.6
```
Comment (v1k0d3n) [inline ansible/roles/kubelet/templates/calico.yaml:188]: could match up with your other version variable for CNI.

File: ansible/roles/kubelet/templates/calico.yaml [hunk 9]
```diff
@@ -0,0 +1,322 @@
...
+        scheduler.alpha.kubernetes.io/tolerations: |
+          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
+           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
+    spec:
+      # The policy controller must run in the host network namespace so that
+      # it isn't governed by policy that would prevent it from working.
+      hostNetwork: true
+      containers:
+        - name: calico-policy-controller
+          image: calico/kube-policy-controller:v0.5.2
```
Comment (v1k0d3n) [inline ansible/roles/kubelet/templates/calico.yaml:257]: same with versioning.

File: ansible/roles/kubelet/templates/calico.yaml [hunk 10]
```diff
@@ -0,0 +1,322 @@
...
+        scheduler.alpha.kubernetes.io/tolerations: |
+          [{"key": "dedicated", "value": "master", "effect": "NoSchedule" },
+           {"key":"CriticalAddonsOnly", "operator":"Exists"}]
+    spec:
+      hostNetwork: true
+      restartPolicy: OnFailure
+      containers:
+        # Writes basic configuration to datastore.
+        - name: configure-calico
+          image: calico/ctl:v1.0.2
```
Comment (v1k0d3n) [inline ansible/roles/kubelet/templates/calico.yaml:301]: same, if the others are versioned.

**Comment by v1k0d3n** [comment]
@aric49, awesome work! this is really coming together nicely. just a few thoughts.

**Comment by aric49** [comment]
Just wanted to check in my latest changes so they don't get lost and are out there as a work in progress. Your suggestions @v1k0d3n are still noted, and will be implemented once I am able to get a stable code base :)

**Comment by v1k0d3n** [comment]
are these changes resulting in a successful cluster? i thought there were still some issues, which is why i held back on pulling them in.

**Comment by aric49** [comment]
Yes, that's correct. Still a lot of issues going on.  However, i wanted to check in my latest changes so anyone  can look and provide input on them (also, so they don't get lost).   The current issues I'm seeing are on the bootkube issues page: https://github.com/kubernetes-incubator/bootkube/issues/407

**Comment by v1k0d3n** [comment]
ok, so agreement is that i will start picking this up next week and report any new issues that could creep up. @aric49 if you can write the once you currently know of for tracking purposes...i will consider this an exchange for pulling in what you have.

**Pull request merged**